jetty-websocket/jetty-websocket-tests/src/test/java/org/eclipse/jetty/websocket/tests/WebSocketOverHTTP2Test.java

```diff
@@ -0,0 +1,129 @@
+
+public class WebSocketOverHTTP2Test
+{
+    private Server server;
+    private ServerConnector connector;
+
+    @BeforeEach
+    public void startServer() throws Exception
+    {
+        QueuedThreadPool serverThreads = new QueuedThreadPool();
+        serverThreads.setName("server");
+        server = new Server(serverThreads);
+        HttpConfiguration httpConfiguration = new HttpConfiguration();
+        HttpConnectionFactory h1 = new HttpConnectionFactory(httpConfiguration);
+        HTTP2CServerConnectionFactory h2c = new HTTP2CServerConnectionFactory(httpConfiguration);
+        connector = new ServerConnector(server, 1, 1, h1, h2c);
+        server.addConnector(connector);
+
+        ServletContextHandler context = new ServletContextHandler(server, "/");
+        context.addServlet(new ServletHolder(new JettyWebSocketServlet()
+        {
+            @Override
+            protected void configure(JettyWebSocketServletFactory factory)
+            {
+                factory.addMapping("/ws/echo", (req, resp) -> new EchoSocket());
+            }
+        }), "/ws/*");
+        JettyWebSocketServletContainerInitializer.initialize(context);
+
+        server.start();
+    }
+
+    @AfterEach
+    public void stopServer() throws Exception
+    {
+        server.stop();
+    }
+
+    @Test
+    public void testWebSocketOverDynamicHTTP1() throws Exception
+    {
+        testWebSocketOverDynamicTransport(clientConnector -> HttpClientConnectionFactory.HTTP11);
+    }
+
+    @Test
+    public void testWebSocketOverDynamicHTTP2() throws Exception
+    {
+        testWebSocketOverDynamicTransport(clientConnector -> new ClientConnectionFactoryOverHTTP2.H2C(new HTTP2Client(clientConnector)));
+    }
+
+    private void testWebSocketOverDynamicTransport(Function<ClientConnector, ClientConnectionFactory.Info> protocolFn) throws Exception
+    {
+        ClientConnector clientConnector = new ClientConnector();
+        clientConnector.setSelectors(1);
+        QueuedThreadPool clientThreads = new QueuedThreadPool();
+        clientThreads.setName("client");
+        clientConnector.setExecutor(clientThreads);
+        HttpClient httpClient = new HttpClient(new HttpClientTransportDynamic(clientConnector, protocolFn.apply(clientConnector)));
+
+        WebSocketClient wsClient = new WebSocketClient(httpClient);
+        wsClient.start();
+
+        EventSocket wsEndPoint = new EventSocket();
+        URI uri = URI.create("ws://localhost:" + connector.getLocalPort() + "/ws/echo");
+        Session session = wsClient.connect(wsEndPoint, uri).get(5, TimeUnit.SECONDS);
+
+        String text = "http2";
+        session.getRemote().sendString(text);
+
+        String message = wsEndPoint.messageQueue.poll(5, TimeUnit.SECONDS);
+        assertNotNull(message);
+        assertEquals(text, message);
+
+        session.close(StatusCode.NORMAL, null);
+        assertTrue(wsEndPoint.closeLatch.await(5, TimeUnit.SECONDS));
```

Comment:
should check after this that the reason of closed was `StatusCode.NORMAL` and perhaps that no error occurred

Reply:
Done.